backends/tfhe-cuda-backend/cuda/src/integer/compression/compression.cuh

```diff
@@ -0,0 +1,226 @@
+
+    host_fp_keyswitch_lwe_list_to_glwe(
+        streams[0], gpu_indexes[0], glwe_out, lwe_subset, fp_ksk[0],
+        fp_ks_buffer, input_lwe_dimension, compression_params.glwe_dimension,
+        compression_params.polynomial_size, compression_params.ks_base_log,
+        compression_params.ks_level, min(num_lwes, mem_ptr->lwe_per_glwe));
+  }
+
+  auto body_count = min(num_lwes, mem_ptr->lwe_per_glwe);
+
+  // Modulus switch
+  modulus_switch_inplace(streams[0], gpu_indexes[0], tmp_glwe_array_out,
+                         num_glwes * (compression_params.glwe_dimension *
+                                          compression_params.polynomial_size +
+                                      body_count),
+                         mem_ptr->storage_log_modulus);
+  check_cuda_error(cudaGetLastError());
+
+  pack(streams[0], glwe_array_out, tmp_glwe_array_out, num_glwes, body_count,
+       mem_ptr);
+}
+
+template <typename Torus>
+__global__ void kernel_extract(Torus *glwe_array_out, Torus *array_in,
+                               uint32_t index, uint32_t log_modulus,
+                               uint32_t initial_out_len) {
+  auto nbits = sizeof(Torus) * 8;
+
+  auto i = threadIdx.x + blockIdx.x * blockDim.x;
+
+  if (i < initial_out_len) {
+    // Unpack
+    Torus mask = ((Torus)1 << log_modulus) - 1;
+    auto start = i * log_modulus;
+    auto end = (i + 1) * log_modulus;
+
+    auto start_block = start / nbits;
+    auto start_remainder = start % nbits;
+
+    auto end_block_inclusive = (end - 1) / nbits;
+
+    Torus unpacked_i;
+    if (start_block == end_block_inclusive) {
+      auto single_part = array_in[start_block] >> start_remainder;
+      unpacked_i = single_part & mask;
+    } else {
+      auto first_part = array_in[start_block] >> start_remainder;
+      auto second_part = array_in[start_block + 1] << (nbits - start_remainder);
+
+      unpacked_i = (first_part | second_part) & mask;
+    }
+
+    // Extract
+    glwe_array_out[i] = unpacked_i << (nbits - log_modulus);
+  }
+}
+
+template <typename Torus>
+__host__ void extract(cudaStream_t stream, uint32_t gpu_index,
+                      Torus *glwe_array_out, Torus *array_in,
+                      uint32_t glwe_index, int_compression<Torus> *mem_ptr) {
+
+  auto params = mem_ptr->compression_params;
+
+  auto log_modulus = mem_ptr->storage_log_modulus;
+
+  uint32_t body_count = mem_ptr->body_count;
+  auto initial_out_len =
+      params.glwe_dimension * params.polynomial_size + body_count * body_count;
+
+  // We assure the tail of the glwe is zeroed
+  auto zeroed_slice = glwe_array_out + initial_out_len;
+  cuda_memset_async(zeroed_slice, 0,
+                    (params.polynomial_size - body_count * body_count) *
+                        sizeof(Torus),
+                    stream, gpu_index);
+
+  int num_blocks = 0, num_threads = 0;
+  getNumBlocksAndThreads(initial_out_len, 512, num_blocks, num_threads);
```

Comment:
Whhy 512 threads here? Have you tried with 1024?

Reply:
No strong reason for 512. Just changed to 1024 and benchmarks are a bit worst. Actually, 128 gave me the best result.